src/coreclr/jit/flowgraph.cpp

```diff
@@ -4184,3 +4191,321 @@ void Compiler::fgLclFldAssign(unsigned lclNum)
+{
+    PhaseStatus result = PhaseStatus::MODIFIED_NOTHING;
+
+    if (!doesMethodHaveTlsFieldAccess())
+    {
+        // The method doesn't have any TLS field
+        return result;
+    }
+
+    CORINFO_THREAD_STATIC_BLOCKS_INFO threadStaticBlocksInfo;
+    info.compCompHnd->getThreadLocalStaticBlocksInfo(&threadStaticBlocksInfo);
+    assert(threadStaticBlocksInfo.tlsIndex.accessType == IAT_VALUE);
+
+    for (BasicBlock* block = fgFirstBB; block != nullptr; block = block->bbNext)
+    {
+    SCAN_BLOCK_AGAIN:
+        for (Statement* const stmt : block->Statements())
+        {
+            if ((stmt->GetRootNode()->gtFlags & GTF_CALL) == 0)
+            {
+                // TP: Stmt has no calls - bail out
+                continue;
+            }
+
+            for (GenTree* const tree : stmt->TreeList())
+            {
+                // We only need calls with IsExpRuntimeLookup() flag
+                if (!tree->IsCall() || !tree->AsCall()->IsHelperCall())
+                {
+                    continue;
+                }
+                GenTreeCall* call = tree->AsCall();
+
+                if (!call->IsHelperCall())
+                {
+                    continue;
+                }
+                CorInfoHelpFunc func = eeGetHelperNum(call->gtCallMethHnd);
+                if (func != CORINFO_HELP_GETSHARED_NONGCTHREADSTATIC_BASE_NOCTOR_OPTIMIZED)
+                {
+                    continue;
+                }
+
+                JITDUMP("Expanding thread static local access for [%06d] in " FMT_BB ":\n", dspTreeID(tree),
+                        block->bbNum);
+                DISPTREE(tree);
+                JITDUMP("\n");
+
+                assert(call->gtArgs.CountArgs() == 1);
+
+                // Split block right before the call tree
+                BasicBlock* prevBb       = block;
+                GenTree**   callUse      = nullptr;
+                Statement*  newFirstStmt = nullptr;
+                DebugInfo   debugInfo    = stmt->GetDebugInfo();
+                block                    = fgSplitBlockBeforeTree(block, stmt, call, &newFirstStmt, &callUse);
+                assert(prevBb != nullptr && block != nullptr);
+
+                // Block ops inserted by the split need to be morphed here since we are after morph.
+                // We cannot morph stmt yet as we may modify it further below, and the morphing
+                // could invalidate callUse.
+                while ((newFirstStmt != nullptr) && (newFirstStmt != stmt))
+                {
+                    fgMorphStmtBlockOps(block, newFirstStmt);
+                    newFirstStmt = newFirstStmt->GetNextStmt();
+                }
+
+                GenTreeLclVar* threadStaticBlockLcl = nullptr;
+
+                // Mostly for Tier0: if the current statement is ASG(LCL, RuntimeLookup)
+                // we can drop it and use that LCL as the destination
+                if (stmt->GetRootNode()->OperIs(GT_ASG))
+                {
```

Comment:
I assume you can remove the whole block here, it was some hack to mitigate tier0 regressions in runtime lookups

Reply:
thanks for confirming.